src/data/set/ncard.lean

```diff
@@ -0,0 +1,630 @@
+
+/-
+Copyright (c) 2023 Peter Nelson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Peter Nelson
+-/
+import data.finite.card
+
+/-!
+# Noncomputable Set Cardinality
+
+We define the cardinality `set.ncard s` of a set `s` as a natural number. This function is
+noncomputable (being defined in terms of `nat.card`) and takes the value `0` if `s` is infinite.
+
+This is intended as an alternative to `finset.card` and `fintype.card`, both of which contain data
+in their definition that can cause awkwardness when using `set.to_finset`. Using `set.ncard` allows
+cardinality computations to avoid `finset`/`fintype` completely, staying in `set` and letting
+finiteness be handled either explicitly or via typeclasses. An ideal use case is when working
+with terms in `set α` where a `finite α` instance is present.
+
+## Main Definitions
+
+* `set.ncard s` is the cardinality of the set `s` as a natural number, provided `s` is finite.
+  If `s` is infinite, then `set.ncard s = 0`.
+
+## Implementation Notes
+
+The lemmas in this file are very similar to those in `data.finset.card`, but with `set` operations
+instead of `finset`; most of the proofs invoke their `finset` analogues. Nearly all the lemmas
+require finiteness of one or more of their arguments, which we give as `[finite s]`. Often, where
+there are two arguments `s` and `t`, the finiteness of one follows from the other in the context
+of the lemma, in which case we only include the ones that are needed, and derive the other inside
+the proof.
+-/
+
+open_locale classical
+
+variables {α β : Type*} {s t : set α} {a b x y : α} {f : α → β}
+
+namespace set
+
+/-- The cardinality of `s : set α` . Has the junk value `0` if `s` is infinite -/
+noncomputable def ncard (s : set α) := nat.card s
+
+lemma ncard_def (s : set α) : s.ncard = nat.card s := rfl
+
+lemma ncard_eq_to_finset_card (s : set α) [finite s] :
```

Comment:
I think the lemmas in this file should use `s.finite` not `finite s`; `set.finite` is to `finite` as `set.ncard` is to `nat.card`.

Reply:
The disadvantage of this is that is makes the primary use case I had in mind (sets in a `finite` type) a lot more awkward.

Reply:
You can use `hs : s.finite . mytac`, where `mytac` is defined as `apply set.to_finite`. The normal form for saying "a set is finite" is `set.finite s`, not `[finite s]` (unless you're defining an instance).

Reply:
Would you mind showing me the exact code used to define `mytac` in this way? I have never done any metaprogramming.